### ISSUE TITLE
Scylla IPV6 - seed nodes do not connect to peers with scope

### ISSUE THREAD

**Issue by slivne** [scylladb/scylla#5225]
Scylla: 666.development-0.20191025.8bffb800e19

I have setup a 2 node cluster using the AMI: scylla_2019-10-26T04-01-45Z (ami-0ffd133300b2d509e)

Seed node setup
```
[centos@ip-10-0-0-225 ~]$ cat /etc/scylla/scylla.yaml  | grep fe80
          - seeds: "fe80::852:7cff:fec4:7be1%2"
listen_address: fe80::852:7cff:fec4:7be1%2
rpc_address: fe80::852:7cff:fec4:7be1%2
broadcast_rpc_address: fe80::852:7cff:fec4:7be1%2

+
enable_ipv6_dns_lookup: true
```

I have also followed QA instructions on updating

/etc/sysconfig/network-scripts/ifcfg-eth0
```
IPV6_AUTOCONF=yes
IPV6_DEFROUTE=yes
```

and restarting networking
```
sudo systemctl restart network
```

I did not define the route

The seed node boots ok and when I do netstat -an I can see the binding is set
```
netstat -an | grep 7000
tcp6       0      0 fe80::852:7cff:fec:7000 :::*                    LISTEN     
```

The second node is setup with 
```
[centos@ip-10-0-0-188 ~]$ cat /etc/scylla/scylla.yaml  | grep fe80
          - seeds: "fe80::852:7cff:fec4:7be1%2"
listen_address: fe80::8e8:e0ff:fe6e:5c3d%2
rpc_address: fe80::8e8:e0ff:fe6e:5c3d%2
broadcast_rpc_address: fe80::8e8:e0ff:fe6e:5c3d%2
+
enable_ipv6_dns_lookup: true
```

When the node2 connects it seems to be able to connect to the seed node

that is on the seed node

```
 netstat -an | grep 7000
tcp6       0      0 fe80::852:7cff:fec:7000 :::*                    LISTEN     
tcp6       0      0 fe80::852:7cff:fec:7000 fe80::8e8:e0ff:fe:61474 ESTABLISHED
```

Yet the seed node is not able to connect back to node2, strace on the seed node I get

```
socket(AF_INET6, SOCK_STREAM|SOCK_CLOEXEC|SOCK_NONBLOCK, IPPROTO_TCP) = 36
setsockopt(36, SOL_SOCKET, SO_REUSEADDR, [1], 4) = 0
setsockopt(36, SOL_IP, 0x18 /* IP_??? */, [1], 4) = 0
bind(36, {sa_family=AF_INET6, sin6_port=htons(59916), inet_pton(AF_INET6, "::", &sin6_addr), sin6_flowinfo=0, sin6_scope_id=4294967295}, 28) = 0
connect(36, {sa_family=AF_INET6, sin6_port=htons(7000), inet_pton(AF_INET6, "fe80::8e8:e0ff:fe6e:5c3d", &sin6_addr), sin6_flowinfo=0, sin6_scope_id=4294967295}, 28) = -1 ENETUNREACH (Network is unreachable)
sendto(11, "<14>Oct 26 23:14:40 scylla:  [sh"..., 134, MSG_NOSIGNAL, NULL, 0) = 134
```

journalctl has

```
Oct 26 23:18:05 ip-10-0-0-225.ec2.internal scylla[4466]:  [shard 0] gossip - Fail to handle GOSSIP_DIGEST_SYN: seastar::rpc::closed_error (connection is closed)
Oct 26 23:18:05 ip-10-0-0-225.ec2.internal scylla[4466]:  [shard 0] rpc - client [fe80::8e8:e0ff:fe6e:5c3d]:7000: fail to connect: connect: Network is unreachable
Oct 26 23:18:04 ip-10-0-0-225.ec2.internal scylla[4466]:  [shard 0] gossip - Fail to handle GOSSIP_DIGEST_SYN: seastar::rpc::closed_error (connection is closed)
Oct 26 23:18:04 ip-10-0-0-225.ec2.internal scylla[4466]:  [shard 0] rpc - client [fe80::8e8:e0ff:fe6e:5c3d]:7000: fail to connect: connect: Network is unreachable
```

if i ping6 the address of node2 from the seed it works

```
[centos@ip-10-0-0-225 ~]$ ping6 fe80::8e8:e0ff:fe6e:5c3d%2
PING fe80::8e8:e0ff:fe6e:5c3d%2(fe80::8e8:e0ff:fe6e:5c3d%eth0) 56 data bytes
64 bytes from fe80::8e8:e0ff:fe6e:5c3d%eth0: icmp_seq=1 ttl=64 time=0.270 ms
64 bytes from fe80::8e8:e0ff:fe6e:5c3d%eth0: icmp_seq=2 ttl=64 time=0.293 ms
64 bytes from fe80::8e8:e0ff:fe6e:5c3d%eth0: icmp_seq=3 ttl=64 time=0.283 ms
```  

I also verified telnet from seed to node2

```
[centos@ip-10-0-0-225 ~]$ telnet fe80::8e8:e0ff:fe6e:5c3d%2 7000
Trying fe80::8e8:e0ff:fe6e:5c3d%eth0...
Connected to fe80::8e8:e0ff:fe6e:5c3d%2.
Escape character is '^]'.
^C^]
```

so routing from seed to node2 seems ok 

**It seems the "scope" of node2 is not "saved"**

**Comment by elcallio** [comment]
Den 2019-10-27 kl. 01:30, skrev Shlomi Livne:
>
> Scylla: 666.development-0.20191025.8bffb800e19
>
> I have setup a 2 node cluster using the AMI: 
> scylla_2019-10-26T04-01-45Z (ami-0ffd133300b2d509e)
>
> Seed node setup
>
> |[centos@ip-10-0-0-225 ~]$ cat /etc/scylla/scylla.yaml | grep fe80 - 
> seeds: "fe80::852:7cff:fec4:7be1%2" listen_address: 
> fe80::852:7cff:fec4:7be1%2 rpc_address: fe80::852:7cff:fec4:7be1%2 
> broadcast_rpc_address: fe80::852:7cff:fec4:7be1%2 + 
> enable_ipv6_dns_lookup: true |
>
> I have also followed QA instructions on updating
>
> /etc/sysconfig/network-scripts/ifcfg-eth0
>
> |IPV6_AUTOCONF=yes IPV6_DEFROUTE=yes |
>
> and restarting networking
>
> |sudo systemctl restart network |
>
> I did not define the route
>
> The seed node boots ok and when I do netstat -an I can see the binding 
> is set
>
> |netstat -an | grep 7000 tcp6 0 0 fe80::852:7cff:fec:7000 :::* LISTEN |
>
> The second node is setup with
>
> |[centos@ip-10-0-0-188 ~]$ cat /etc/scylla/scylla.yaml | grep fe80 - 
> seeds: "fe80::852:7cff:fec4:7be1%2" listen_address: 
> fe80::8e8:e0ff:fe6e:5c3d%2 rpc_address: fe80::8e8:e0ff:fe6e:5c3d%2 
> broadcast_rpc_address: fe80::8e8:e0ff:fe6e:5c3d%2 + 
> enable_ipv6_dns_lookup: true |
>
> When the node2 connects it seems to be able to connect to the seed node
>
> that is on the seed node
>
> |netstat -an | grep 7000 tcp6 0 0 fe80::852:7cff:fec:7000 :::* LISTEN 
> tcp6 0 0 fe80::852:7cff:fec:7000 fe80::8e8:e0ff:fe:61474 ESTABLISHED |
>
> Yet the seed node is not able to connect back to node2, strace on the 
> seed node I get
>
> |socket(AF_INET6, SOCK_STREAM|SOCK_CLOEXEC|SOCK_NONBLOCK, IPPROTO_TCP) 
> = 36 setsockopt(36, SOL_SOCKET, SO_REUSEADDR, [1], 4) = 0 
> setsockopt(36, SOL_IP, 0x18 /* IP_??? */, [1], 4) = 0 bind(36, 
> {sa_family=AF_INET6, sin6_port=htons(59916), inet_pton(AF_INET6, "::", 
> &sin6_addr), sin6_flowinfo=0, sin6_scope_id=4294967295}, 28) = 0 
> connect(36, {sa_family=AF_INET6, sin6_port=htons(7000), 
> inet_pton(AF_INET6, "fe80::8e8:e0ff:fe6e:5c3d", &sin6_addr), 
> sin6_flowinfo=0, sin6_scope_id=4294967295}, 28) = -1 ENETUNREACH 
> (Network is unreachable) |
You can see here that the scope of the address got lost somewhere. (It 
is -1u)

**Comment by elcallio** [comment]
So the problem is probably this: When we communicate across nodes, for gossip, the connecting non-seed node sends his own address to the seed. When serializing ip addresses, we don't include scope, so when the seed connects back, he will use default (-1) scope -> fail. 

However, serializing scope with address would solve nothing, since scope is _machine_ local (designates an interface). And the scope to connect back to a link-local address on a different machine is probably gonna be different (although probably not on AMI:s, but they are clones and do not count in the argument). 

I think the "correct" way to handle this is to more or less do what ping etc does when given a link-local address: Guess. I.e. look at available interfaces, and guess which one will be best route for this address. 
Using linux getaddrinfo is probably safest way to do so. Not sure how blocking this is however...

**Comment by elcallio** [comment]
In origin this problem is ignored, because apparently, if scope is not set, Java tries to figure out the correct scope by querying routing. 
Which of course, again, could fail spectacularly. But maybe we need to add this to seastar?

**Comment by elcallio** [comment]
ping @gleb-cloudius  @asias

**Comment by elcallio** [comment]
Fwiw, I sent an RFC patch doing pretty much what java does to the seastar list.

**Comment by gleb-cloudius** [comment]
On Mon, Oct 28, 2019 at 07:54:01AM -0700, Calle Wilund wrote:
> * Node A binds to a link-local address, connects to seed B
> * B get an accept -> connected socket. Note that this connection originates on whatever A:s connect bound the connecting ::0:<port>to, which is probably _not_ going to be the link-local address he is listening to. 
> * The rpc message goes to gossip. Gossip gets A's broadcast address from extra info in rpc clientinfo, i.e. serialized inet_address. 
Can we get an interface the connection receives packets from and put it
into clientinfo as well? Theoretically different packets may come from
different interfaces, but it should not be the case for link-local
addressing.

--
			Gleb.

**Comment by elcallio** [comment]
Well, sure we can. But that would still only help with replying to an incoming message. But (afaik) we also have the case of trying to connect to addresses retrieved by SYN or loaded from db storage. And those we have no real way of determining interface routing for.  Other than maybe doing (either automatic, like in patch I sent, or explicit) route introspection and trying to figure out where we should send messages.

**Comment by gleb-cloudius** [comment]
Whoever invented IPV6 scopes did not think it through it seams.

--
			Gleb.

**Comment by slivne** [comment]
@elcallio merged a fix into seastar we need to see its merged to scylla and then check

scylladb/seastar@75e189c6ba5c4a9e17432932474fac27ae2fff66

**Comment by elcallio** [comment]
So there still remains a few issues.
1.) (seastar bug-ish) std::hash for inet_address is inconsistent with operator==. == says true if actual address and scope is equal, while hash ignores the scope and just hashes the address part. Not actually wrong given the reqs for hash, but not great either. 
2.) Real culprit: In scylla rpc serialization of inet_addresses, we only send the address part. This is technically wholly correct, as the scope (i.e. interface part) has zero meaning on a different machine. 
However, since endpoints are hashed on endpoint address, having a (seed) node with a link-local address will foul this up. 
I think the real fix is to make gossip endpoint map use address explicitly separated from scope in the map, i.e. filter out any scope when adding itself.

**Comment by elcallio** [comment]
See https://github.com/scylladb/seastar/issues/704

**Comment by slivne** [comment]
the fix for scylladb/seastar#704 was merged 

scylladb/seastar@cfb20c519b57ff61bfcf690f16654cc4cc7e0d15

```
commit cfb20c519b57ff61bfcf690f16654cc4cc7e0d15
Author: Calle Wilund <calle@scylladb.com>
Date:   Wed Nov 20 15:58:49 2019 +0000

    inet_address: Make inet_address == operator ignore scope (again)
    
    Fixes #704
    
    Because of how inet_address:es are typically used in client apps (esp.
    certain database apps), to achieve full transparent behaviour, we
    should ignore the scope part of the address when doing comparison.
    
    This is all based in apps doing serialization of addresses, and
    comparing them on different hosts. Note that adding scopes to
    serialization in these cases are not an options, since the scope
    is not relevant on a different host, thus should be considered
    serialization transient. However, same serialized addresses are
    compared to "local" versions of the same, but with scope in them.
    And those compare as different. We don't want that. So should make
    comparison ignore scope, and treat this as wholly local/transient.
    Yet propagate it so bind etc works.
    
    This is all somewhat despicable behaviour, but consistent with the
    almighty jdk forefather, where any address is considered equal iff
    the bytes comprising it are same length and same content. And scope
    is not part of the bytes.
    
    Message-Id: <20191120155849.25206-1-calle@scylladb.com>
```

@avikivity can we backport the seastar fix for IPV6

**Comment by avikivity** [comment]
Deferring backport until after 3.2 release.

**Comment by avikivity** [comment]
Fixed by 2b7e97514a658ddd406aa13da6df495c1bfee8ea

**Comment by avikivity** [comment]
Backported to 3.2.
